ucp/_libs/ucx_api.pyx

```diff
@@ -446,11 +446,88 @@ cdef class UCXListener(UCXObject):
         return int(<uintptr_t>self._handle)
 
 
+# Counter used as UCXRequest UIDs
+cdef int _ucx_request_counter = 0
+
+cdef class UCXRequest:
+    """Python wrapper of UCX request handle.
+
+    Notice, this class doesn't own the handle and multiple instances of
+    UCXRequest can point to the same underlying UCX handle.
+    Furthermore, UCX can modify/free the UCX handle without notice
+    thus we use `_uid` to make sure the handle hasn't been modified.
+    """
+    cdef:
+        ucp_request *_handle
+        int _uid
+
+    def __init__(self, uintptr_t req_as_int):
+        global _ucx_request_counter
+        cdef ucp_request *req = <ucp_request*>req_as_int
+        assert req != NULL
```

Comment:
Should we raise an error here (perhaps `ValueError`)?

Reply:
I am not sure, this is really an internal check. 
This only fails if UCX calls a callback function with a NULL argument, which mean something has gone seriously wrong :)

Reply:
That's fair. Leaving it makes sense then.